### PR TITLE
Implement reflective support for Java Records

### LINKED ISSUE #1278: Support Java record classes when they are stable
As of the latest release `2.11.0` (and probably past releases too) Moshi is not able to deserialize Java Records, it simply fails with an `AssertionError` when calling `Adapter#fromJson` with an appropriate JSON object.
See the attached very simple sample project for more details: [test_project.zip](https://github.com/square/moshi/files/5630567/test_project.zip)
Stacktrace:
```
Exception in thread "main" java.lang.AssertionError
	at com.squareup.moshi.ClassJsonAdapter.fromJson(ClassJsonAdapter.java:208)
	at com.squareup.moshi.internal.NullSafeJsonAdapter.fromJson(NullSafeJsonAdapter.java:41)
	at com.squareup.moshi.JsonAdapter.fromJson(JsonAdapter.java:51)
	at Main.main(Main.java:8)
```

### PR DESCRIPTION
**Pull request by ZacSweers**
Porting from https://github.com/ZacSweers/MoshiX/tree/main/moshi-records-reflect

This makes a few changes along the way to make the build all happy with how we're doing this:
- Standardizing around Java 8 (which is the minimum supported by gradle toolchains and also soon to be the minimum supported by Kotlin)
- Create a separate Java 16 source set for hosting the actual adapter implementation. This is compiled with Java 16 but is then packaged into the main Moshi jar and denoted as a Multi-Release JAR.
- We gate access to this record adapter at runtime by checking if the requisite record APIs are available.
- We separately have a shim copy of the record adapter in the `main` source set for linking in `BUILT_IN_FACTORIES`, but this is booted out by the real one during packaging.
- Replace animal sniffer with `-release 8`, which functionally accomplishes the same
- Tests are in a dedicated separate `records-test` subproject, making it easier to both write new ones and also ensure our packaging rules for the jar is working correctly (we've had success with this in the past with the codegen artifact's shading being tested in a dedicated tests project)

Resolves #1278

Tests are a little skimpy right now, happy to take suggestions on more we should add! Can also look at fleshing them out more in a followup PR

### CODE REVIEW

**Comment by ZacSweers** [comment]
It looks like animal sniffer ships a shaded old version of ASM and is also no longer being updated since the `-release` flag covers the same on JDK 9+. Thing is we want to support 8, so we can't use it yet. Seems like we'll need to drop animal sniffer if we want to proceed. Thoughts?

File: moshi/build.gradle.kts [hunk 1]
```diff
@@ -22,11 +22,43 @@ plugins {
...
+  options.release.set(16)
+}
+
+// We need to include our actual RecordJsonAdapter from java16 sources and replace the shime
+tasks.named<Jar>("jar") {
+  from(java16.output) {
+    duplicatesStrategy = DuplicatesStrategy.INCLUDE
+  }
+  manifest {
+    // Indicate that this is a multi release jar
```
Comment (melix) [inline moshi/build.gradle.kts:47]: Is it? Just looks like a regular jar to me.

Comment (ZacSweers) [reply moshi/build.gradle.kts:47]: I was sort of assuming this was needed if the jar contains classes targeting multiple java versions but can omit it if it doesn't really change anything

Comment (ZacSweers) [reply moshi/build.gradle.kts]: 3eb768f

File: moshi/src/main/java/com/squareup/moshi/internal/Util.java [hunk 2]
```diff
@@ -55,6 +56,8 @@
   /** A map from primitive types to their corresponding wrapper types. */
   private static final Map<Class<?>, Class<?>> PRIMITIVE_TO_WRAPPER_TYPE;
 
+  private static final AtomicBoolean RECORDS_AVAILABLE = new AtomicBoolean(false);
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/internal/Util.java:59]: Normal `boolean`? Static init blocks cannot run concurrently.

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                + replaced.name
+                + "\n"
+                + "    "
+                + componentBinding.name);
+      }
+    }
+    Constructor<Object> constructor;
+    try {
+      //noinspection unchecked
+      constructor = (Constructor<Object>) rawType.getDeclaredConstructor(constructorParams);
```
Comment (JakeWharton) [inline moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:115]: Is there no way to get a MethodHandle from the Class once we know it's a record? Can't believe they expect you to use ancient java.lang.reflect reflection to programmatically instantiate these...

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    } catch (NoSuchMethodException e) {
+      throw new AssertionError(e);
+    }
+    return new RecordJsonAdapter<>(constructor, bindings).nullSafe();
+  };
+
+  private static record ComponentBinding<T>(String name, String jsonName, JsonAdapter<T> adapter, Method accessor) {
+  }
+
+    private final String targetClass;
```
Comment (JakeWharton) [inline moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:126]: Wrong indent. GJF bug?

File: build.gradle.kts [hunk 5]
```diff
@@ -70,7 +69,7 @@ spotless {
       "**/TypesTest.java"
     )
     val configureCommonJavaFormat: JavaExtension.() -> Unit = {
-      googleJavaFormat("1.7")
+      googleJavaFormat("1.10")
```
Comment (JakeWharton) [inline build.gradle.kts:72]: 1.11 latest

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+    @Override
+    public void toJson(JsonWriter writer, T value) throws IOException {
+      writer.beginObject();
+
+      for (var binding : componentBindingsArray) {
+        writer.name(binding.jsonName);
+        try {
+          binding.adapter.toJson(writer, binding.accessor.invoke(value));
+        } catch (IllegalAccessException | InvocationTargetException e) {
```
Comment (JakeWharton) [inline moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:181]: `InvocationTargetException` should be unwrapped and re-thrown since it indicates an exception on the other side of the reflection.
```java
catch (InvocationTargetException e) {
  Throwable cause = e.getCause();
  if (cause instanceof RuntimeException) throw (RuntimeException) cause;
  if (cause instanceof Error) throw (Error) cause;
  throw new RuntimeException(cause);
}
```

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+      }
+      reader.endObject();
+
+      var initialAccess = constructor.canAccess(null);
+      try {
+        if (!initialAccess) {
+          constructor.setAccessible(true);
+        }
+        return constructor.newInstance(resultsArray);
+      } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
```
Comment (JakeWharton) [inline moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:166]: Same InvocationTargetException unwrapping

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        throw new AssertionError(e);
+      } finally {
+        constructor.setAccessible(false);
```
Comment (JakeWharton) [inline moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:169]: Is there a reason to reset this? I expect the `canAccess` check to not be cheap such that I'm wondering why we don't make it accessible once and never again.

**Review by JakeWharton — COMMENTED** [review]
Where do you actually pack the MR jar? I don't see any references to `META-INF/versions/16/`

**Comment by ZacSweers** [comment]
> Where do you actually pack the MR jar? I don't see any references to META-INF/versions/16/

Yup removed! I misunderstood the requirements for this

File: moshi/src/main/java/com/squareup/moshi/internal/Util.java (hunk 2 above)
Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/internal/Util.java:59]: I might be missing something but java doesn't seem to like this
image

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 4 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:126]: GJF bug, I'll file something over there

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 3 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:115]: Honestly never used that API, let me look

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 4 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:126]: Actually I think spotless is ignoring this? Let me look more

Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:126]: Yeah spotless is ignoring this. Formatting it with GJF manually for now

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 8 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:169]: Fair enough, I meant to clean this up with the previous cleanup I did with Method

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 3 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:115]: Good news is we can use MethodHandle for this but bad news is `RecordComponent` itself still returns only `Method` for its accessors :/

File: build.gradle.kts (hunk 5 above)
Comment (ZacSweers) [reply build.gradle.kts]: 20d9715

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 4 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java]: d98ee90

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 6 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java]: 765a642 and also part of later changes

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 7 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java]: Folded into 355422c

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 3 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java]: Constructor bit is in 355422c

I also wrapped the accessor Method with one too, but not sure (lack of familiarity) if this is worthwhile 4700a63. Let me know

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 8 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java]: Cleaned up with the constructor conversion to MethodHandle

**Comment by JakeWharton** [comment]
No MR jar probably means tools like Graal and D8 (and others) will choke processing the jar.

Instead of a gated use and stub you can make the stub return null from a factory method and make the 16 class return a real implementation. Then you unconditionally call the factory and gate behavior on non-null values.

**Comment by ZacSweers** [comment]
ah good idea. Thanks for bearing with me, learning something new with these MR jars!

**Comment by ZacSweers** [comment]
Ok think I've got it working now. This is neat!

image

**Comment by ZacSweers** [comment]
Being reminded how much the java module system sucks when you have to configure stuff

File: moshi/src/main/java/com/squareup/moshi/internal/Util.java (hunk 2 above)
Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/internal/Util.java]: Ended up removing this

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -10,14 +10,7 @@ jobs:
...
         java-version:
-          - 8
-          - 9
-          - 10
-          - 11
-          - 12
-          - 13
-          - 14
-          - 15
+          - 16
```
Comment (swankjesse) [inline .github/workflows/build.yml:13]: Not testing on Java 8 is annoying. Something to look into separately from this change.

File: build.gradle.kts [hunk 10]
```diff
@@ -129,15 +125,14 @@ subprojects {
...
+        languageVersion.set(JavaLanguageVersion.of(16))
+      }
     }
-  }
-
-  pluginManager.withPlugin("ru.vyarus.animalsniffer") {
-    dependencies {
-      "compileOnly"(Dependencies.AnimalSniffer.annotations)
-      "signature"(Dependencies.AnimalSniffer.java7Signature)
+    if (project.name != "records-tests") {
```
Comment (swankjesse) [inline build.gradle.kts:132]: Very nice

File: .github/workflows/build.yml (hunk 9 above)
Comment (ZacSweers) [reply .github/workflows/build.yml:13]: We can make the main tests run with java 8 via Gradle toolchains

File: gradle.properties [hunk 11]
```diff
@@ -15,9 +15,36 @@
 #
 
 # For Dokka https://github.com/Kotlin/dokka/issues/1405
-org.gradle.jvmargs=-XX:MaxMetaspaceSize=512m
+org.gradle.jvmargs=-Xmx2048m -Dfile.encoding=UTF-8 \
+  --add-opens=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED \
```
Comment (swankjesse) [inline gradle.properties:19]: These are ... unfortunate. The Dokka issue above seems unrelated. I think this is the tracking issue.
https://youtrack.jetbrains.com/issue/KT-45545

File: kotlin/codegen/build.gradle.kts [hunk 12]
```diff
@@ -27,18 +27,27 @@ plugins {
...
     )
   }
 }
 
-// To make Gradle happy
-java {
-  sourceCompatibility = JavaVersion.VERSION_1_8
-  targetCompatibility = JavaVersion.VERSION_1_8
+tasks.withType<Test>().configureEach {
+  // For kapt to work with kotlin-compile-testing
```
Comment (swankjesse) [inline kotlin/codegen/build.gradle.kts:38]: Tracking here: https://youtrack.jetbrains.com/issue/KT-45545

File: moshi/build.gradle.kts [hunk 13]
```diff
@@ -19,14 +19,49 @@ import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
 plugins {
   kotlin("jvm")
   id("com.vanniktech.maven.publish")
-  id("ru.vyarus.animalsniffer")
+}
+
+val mainSourceSet by sourceSets.named("main")
+val java16 by sourceSets.creating {
```
Comment (swankjesse) [inline moshi/build.gradle.kts:25]: neat

File: moshi/build.gradle.kts [hunk 14]
```diff
@@ -35,6 +70,8 @@ tasks.withType<KotlinCompile>()
   }
 
 dependencies {
+  // So the j16 source set can "see" main Moshi sources
+  "java16Implementation"(mainSourceSet.output)
```
Comment (swankjesse) [inline moshi/build.gradle.kts:74]: off topic: this syntax offends me so much

File: moshi/records-tests/src/test/java/com/squareup/moshi/records/RecordsTest.java [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+  public void genericBoundedRecord() throws IOException {
+    var adapter =
+        moshi.<GenericBoundedRecord<Integer>>adapter(
+            Types.newParameterizedTypeWithOwner(
+                RecordsTest.class, GenericBoundedRecord.class, Integer.class));
+    assertThat(adapter.fromJson("{\"value\":4}")).isEqualTo(new GenericBoundedRecord<>(4));
+  }
+
+  @Test
+  public void qualifiedValues() throws IOException {
```
Comment (swankjesse) [inline moshi/records-tests/src/test/java/com/squareup/moshi/records/RecordsTest.java:160]: awesome

File: moshi/src/main/java/com/squareup/moshi/RecordJsonAdapter.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+package com.squareup.moshi;
+
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * This is just a simple shim for linking in {@link StandardJsonAdapters} and swapped with a real
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/RecordJsonAdapter.java:25]: I really like how this plugs in

Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/RecordJsonAdapter.java:25]: Feels very reminiscent of expect/actual

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+
+        var rawType = Types.getRawType(type);
+        if (!rawType.isRecord()) {
+          return null;
+        }
+
+        Map<String, Type> mappedTypeArgs = null;
+        if (type instanceof ParameterizedType parameterizedType) {
+          Type[] typeArgs = parameterizedType.getActualTypeArguments();
+          var typeVars = rawType.getTypeParameters();
```
Comment (swankjesse) [inline moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:60]: var! (My first time seeing this in real life)

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+          return null;
+        }
+
+        var rawType = Types.getRawType(type);
+        if (!rawType.isRecord()) {
+          return null;
+        }
+
+        Map<String, Type> mappedTypeArgs = null;
+        if (type instanceof ParameterizedType parameterizedType) {
```
Comment (swankjesse) [inline moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:58]: This reads to me as very similar to `Util.resolve()`. Presumably we can’t use that here because this needs more records stuff? Either way this behavior really wants to be extracted into a separately-testable thing

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+        MethodHandle constructor;
+        try {
+          constructor = lookup.findConstructor(rawType, methodType(void.class, constructorParams));
+        } catch (NoSuchMethodException | IllegalAccessException e) {
+          throw new AssertionError(e);
+        }
+        return new RecordJsonAdapter<>(constructor, rawType.getSimpleName(), bindings).nullSafe();
+      };
+
+  private static record ComponentBinding<T>(
```
Comment (swankjesse) [inline moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:136]: yo dog

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+      String name, String jsonName, JsonAdapter<T> adapter, MethodHandle accessor) {}
+
+  private final String targetClass;
+  private final MethodHandle constructor;
+  private final ComponentBinding<Object>[] componentBindingsArray;
+  private final JsonReader.Options options;
```
Comment (swankjesse) [inline moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:142]: Yay options

**Review by swankjesse — COMMENTED** [review]
I think I’d like to do a follow-up to this to test its internal components, though that shouldn’t block release.

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 18 above)
Comment (swankjesse) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:58]: Fix: https://github.com/square/moshi/pull/1442/files

File: moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java (hunk 19 above)
Comment (ZacSweers) [reply moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java:136]: yeah I think the static is implicit but the IDE didn't give me a warning about it being redundant

File: gradle.properties (hunk 11 above)
Comment (ZacSweers) [reply gradle.properties:19]: The kapt parts are no longer relevant or on `master` now at least with Kotlin 1.6
